quickly.1
.TH Quickly 1 "08/20/2009" "" "Quickly"
.\"Created with GNOME Manpages Editor
.\"http://sourceforge.net/projects/gmanedit2

.SH NAME
Quickly \- create software programs (and other things) quickly

.SH SYNOPSIS
.B quickly
.RI [ options ]
.RI [ template ]
.RI command
.RI [ args ]
.br

.SH DESCRIPTION
This manual page explains the \fBquickly\fP program. The \fBquickly\fP program helps you create software programs (and other things) quickly. You can select from a set of application templates and use some simple quickly commands to create, edit code and GUI, and publish your software for others to use. Quickly's templates are easy to write. So if you are a fan of language foo, you can create a foo-project template. Or if you want to help people making plugins for your killer app, you can make a killer-app-plugin template. You can even create a template for managing corporate documents, creating your own awesome LaTeX helpers. The sky is the limit!
.P
Given the richness and variety of the Linux platform, programming on Linux
should be easy and fun. However, it's not easy and fun because it is too hard
to learn. It is too hard to learn because there are too many choices, and too
much information to wade through. Quickly strives to make programming easy and
fun by bringing opinionated choices about how to write different kinds of
programs to developers. Included is a Ubuntu application template for making applications that integrate smoothly into the Ubuntu software infrastructure.
.P
GTD is easier with GTQD (Get Things Quickly Done)!

.SH OPTIONS
Quickly follows  the usual GNU command line syntax, with long options starting with two dashes (‘-’). A summary of options is included below.

.IP \fB\-h\fP
.IP \fB\--help\fP
Show help information

.IP \fB\-t\fP
.IP \fB\--templates\fP
This option is followed by a template name. It's needed for template commands launched outside a project or can be used in a project for launching commands from another template.

.IP \fB\--staging\fP
Target launchpad staging server. Can automatically be added for using in multiple command suits if export QUICKLY="staging".

.IP \fB\--verbose\fP
All commands are launched in verbose. Can automatically be added for using in multiple command suits if export QUICKLY="verbose".

.IP \fB\--version\fP
Show version information.

.SH COMMANDS

Each command can be launched with or without options.

.TP
.B command args
Launch the command with args in the current contexte (inside or outside a project). Some commands needs to be launched inside a quickly project (for instance, quickly save in ubuntu-application template), some other can be launched only outside a project (quickly create) and some can be launched inside or outside a project (quickly quickly)
.P
If a command is launched inside a project, the current associated template will be automatically the one used at project creation (probably with quickly create template projectname). You can still launch builtin command (commands that doesn't depend on any template) or command from another template with quickly -t anothertemplate command_from_anothertemplate.
.P
Some commands (like quickly create, quickly quickly, quickly help) needs a template after their command name if you aren't in a project directory. This can also be achieved in using the -t template option.

.TP
.B create [ template ] projectname
Create a new project from template. Can only be launched outside an existing project.
.P
You can also use:
.RS
quickly -t template create projectname (equivalent use)
.RE

.IP \fBcommands\fP
List all commands available, sorted by template

.IP \fBgetstarted\fP
Have some info to get started

.TP
.B quickly [ template ] newtemplate
Create a new personal template from an existing template. You can then modify this template for your own needs.
.P
You can also use:
.RS
quickly -t template quickly newtemplate (equivalent use)
.P
or:
.P
quickly quickly newtemplate if you are in a project. The newtemplate will derivate from associated project template.
.RE

.TP
.B help [ template ] command
Get some help on a command. If no template is provided and you are outside any quickly project, you are only able to get some help from builtin commands.
If you specify a template (or use quickly help -t template command, or you are in a quickly project directory), you will get some help from the current template context.

.SH ENVIRONMENT

.TP
.B QUICKLY
The \fBQUICKLY\fP environment variable can contain multiple tags. The separator is \fB":"\fP.

For example: \fBQUICKLY=staging:verbose\fP

.TP
.B QUICKLY_TEMPLATES
This variable contains additional template paths, separated by \fB":"\fP much
like the \fBPATH\fP environment variable.

E.g.: \fBQUICKLY_TEMPLATES=/home/foo/.quickly/templates:/home/foo/projects/mynewtemplate\fP



NEWS
Quickly 11.04
------------------------------------

 + ubuntu-application-template:
   - Separated user-modifiable code from quickly-owned code by adding new
     *_lib module in new projects that contains quickly-maintained code
   - Support self.ui['my crazy widget'] widget access
   - Refresh tutorial for Ubuntu 11.04 with new screenshots and more editing
 + ubuntu-application-template and derivatives:
   - Add 'quickly test' command that will run unit tests on your code; some
     sample tests are included
   - Fixed LP login support to work with latest liblaunchpad
 + ubuntu-pygame-template:
   - Add hiscores support by Alejandro J. Cura

Quickly 11.03.1
------------------------------------

 + ubuntu-application-template:
   - Use one scalable application icon, instead of 3 differently sized ones
 + ubuntu-application-template and derivatives:
   - Support writing metadata required by the Application Review Board when
     packaging (can enable with 'submitubuntu' or 'package --extras')
   - Add yelp to dependencies if your application uses ghelp: URLs
   - Some minor cleanups
 + ubuntu-flash-game-template:
   - First release of a flash game template by Stuart Langridge

Quickly 11.03
------------------------------------

 + ubuntu-application-template:
   - tutorial command reference 4.4 glade should be design (Tony Byrne)
     LP: #661699
   - refresh the branding to the new ubuntu brand (Daniel Fore) (LP: #666557)
   - Support self.ui.label1 syntax
   - If glade defined handler is missing connect to default_handler.
   - Support auto-connection of signal handlers named on_widget_signal()
   - Support auto-connection of several widgets to a handler.
   - support submitubuntu command to install applications in /opt (LP: #625581)
 + ubuntu-application-template and derivatives:
   - Fix apport code to not fail when lp-project name changes
   - Fix apport test to run cleanly by always upgrading from 0.3 to 0.4
     template (test was originally written to test that upgrade)
   - Add many tests
   - Provide mallard-formatted starter help files (Tony Byrne)
   - Fix 'add dialog' to rename dialog name with dashes correctly
   - Add 'add help-guide' and 'add help-topic' (Tony Byrne)
   - Support custom licenses better by noticing when they are being used
   - add QUICKLY_EDITOR variable to override SELECTED_EDITOR or EDITOR
     if we want a dedicated editor for Quickly (Dennis Craven)
   - regenerate debian/copyright at each quickly release/package (LP: #656943)
   - When getting Launchpad credentials, only allow choosing full access
   - Cleanup various help descriptions to be more consistent
 + common:
   - Reorganize tests to be easier to run as a group (./test/run.sh)
   - quickly quickly should remove *.pyc files as commands are imported
     (LP: #658710)
   - If not running under X, use nano instead of gedit as fallback editor

Quickly 0.6.1
------------------------------------

 + ubuntu-application-template and derivatives:
   - don't overwrite files in add command if already exists (Tony Byrne)
     (LP: #645490)
   - fix crash in in help() (Michael Terry) (LP: #649881)
   - force utf8 when creating a gpg key (LP: #640689)
 + common:
   - revert internal bzr calls to use subprocess calls again
     (LP: #648496, #642455)
   - translations update (en_GB, es, fr, uk)
   - urkkk, just saw that the tutorial contains the name and revision number. 
     Update that as a workaround. Should be removed for next release.

Quickly 0.6
------------------------------------

  quickly core:
  - enhance API with real case usage: nautilus extension
  - move bzr call to use internal API (Jelmer Vernooij)
  - support user defined template paths (Barry Warsaw) (LP: #585617)
  - quickly quickly now uses import command functionality from the origin
    template
  - fix copyright file (Kyle Nitzsche) (LP: #610096)
  ubuntu-application-template and derivatives:
  - push the tarball to launchpad instead of the .changes (LP: #594256)
  - move bzr call to use internal API (Jelmer Vernooij)
  - add apache license (Monty Taylor)
  - update and improvment of the tutorial (Ryan Macnish)
  - "$ quickly add indicator" to add indicator support to your project
    (Owais Lone)
  - add yelp dep for tutorial (Umang Varma) (LP: #610026)
  - some fixes to the tutorial (Henrik H.) (LP: #608421)
  - don't add changelog for quickly package command (Umang Varma)
    (LP: #587538)
  - add quickly configure distribution <distro> to release to another
    distribution than your current one (Umang Varma) (LP: #608191)
  - Provide a modeline in generated source (Michael Terry) (LP: #606285)

Quickly 0.4.3
------------------------------------

  quickly core:
  - fix wrong template proposal for commands not followed by template
  ubuntu-application template:
  - change label caption from glade to design (LP: #571409)
  - fix dialog when having - in title name (LP: #578710)
  - fix unicode problems in About dialog (LP: #582584) (Petar Vasić)
  - fix the boiler plate for faulty preferences dialog code (LP: #587090)
    (Nick Veitch)
  ubuntu-application and inherited:
  - fix configure stripping team name for ppa (LP: #587314) (Frederik Elwert)
  - fix not all syntax for dependencies supported (LP: #587520) (Frederik
    Elwert)
  - fix issue and description of release: quickly release <number>
  updated translations

Quickly 0.4.2
------------------------------------

  ubuntu-application template (and inherited):
  - fix verbose check in quicklyutils making release and share fail
    (LP: #570239)
  - updated translations

Quickly 0.4.1
------------------------------------

  ubuntu-application template (and inherited):
  - add more debugging info in --verbose mode for gpg keys
  - remove ~/.selected_editor detection. Introduced confusion for users
    (LP: #565586)
  - better message and user help when ppa not found (LP: #565581)
  - fix again some tutorial issues which had been lost during html -> docbook

quickly.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from builtins import commands, configurationhandler, tools, launchpadaccess
import os
import sys
import subprocess

import gettext
from gettext import gettext as _

def usage():
    print _("""Usage:
    quickly [OPTIONS] command ...
    
Options:
    -t, --template <template>  Template to use if it differs from default
                               project one
                               one used to create the project)
    --staging                  Target launchpad staging server
    --verbose                  Verbose mode
    -h, --help                 Show help information

Commands:
    new <template> <project-name> (template is mandatory for this command)
    push <describe your changes>

Examples:
    quickly new ubuntu-project foobar
    quickly push 'awesome new comment system'
    quickly -t cool-template push 'awesome new comment system'""")

def look_for_commands(template_path=None):
    """ seek for available commands

TEMPLATE direct…
    template_path: where templates are located. None if we want to list all commands for all templates.

    : return tuples with list of available commands and origin (default or template)
    """

    if template_path is not None:
        print _("in progress")


def check_this_command(command_name, template_path, opt_template):
    """ check if the command exist in a template

            For instance, for a command like foo, the inside template foo.py script
            file is preferred to built-in foo() function.
            There can be pre_foo() and post_foo() built-in functions.

            :command_name
            :template_path

            return list of commands ready to be launched
    """

    commands_available = {}

    # check for template command
    command_path = template_path + "/" + command_name + ".py"
    if os.path.exists(command_path):
        commands_available[opt_template] = command_path
    else:
        #check for built-in command
        if hasattr(commands, command_name):
            commands_available[opt_template] = getattr(commands, command_name)
        else:
            print _("ERROR: command '%s' in '%s' not found.") % (command_name, opt_template)
            print _("Aborting")
            exit(1)        

    #check for pre-post built-in commands
    for hook in ("pre", "post"):
        if hasattr(commands, hook + '_' + command_name):
            commands_available[hook] = getattr(commands, hook + '_' + command_name)

    return commands_available


def process_command_line():
    """ Entry point for command line processing

    :return: exit code of quickly command.
    """
    if len(sys.argv) == 1:
        usage()
        return 0

    opt_command = []
    with_explicit_template = False
    opt_has_template = False
    argv = sys.argv
    i = 1

    while i < len(argv):
        arg = argv[i]
        if arg == '--template' or arg == '-t':
            opt_has_template = True
            opt_template = argv[i + 1]
            i += 1
        elif arg == '--staging':
            oldenv = ""
            if os.environ.has_key('QUICKLY'):
                oldenv = os.environ['QUICKLY']
            os.environ['QUICKLY'] = "staging " + oldenv
        elif arg == '--verbose':
            oldenv = ""
            if os.environ.has_key('QUICKLY'):
                oldenv = os.environ['QUICKLY']
            os.environ['QUICKLY'] = "verbose " + oldenv
        elif arg == '--help' or arg == '-h':
            usage()
            return 0
        else:
            if arg == 'new' or arg == 'quickly':
                with_explicit_template = True
            opt_command.append(arg)
        i += 1

    #if processing command with explicit template, template argument and project name
    #must be there (with -t, --template or just following the command): new, quickly

    if with_explicit_template:
        if not opt_has_template:
            if len(opt_command) < 3 or opt_command[1].startswith('-'):
                print _("ERROR: %s command must be followed by a template and destination name" % opt_command[0])
                print _("Aborting")
                usage()
                return 1
            else:
                opt_template = opt_command[1]
                opt_command.remove(opt_command[1])
                opt_has_template = True

        #in every cases, the project name is now in first position.
        project_name = opt_command[1]
    # the config file already exist, load it.
    else:
        configurationhandler.loadConfig()

    #if no template provided, guess it from the current tree
    if not opt_has_template:
        try:
            opt_template = configurationhandler.config['template']
        except KeyError:
            print _("ERROR: No template provided and none found in the current tree. Ensure you " \
                        "don't want to create a new project or that your are in your directory project.")
            print _("Aborting")
            return 1

    # get the template path
    template_path = tools.get_template_directory(opt_template)

    #ensure the command exists
    if not opt_command:
        print _("ERROR: No command found")
        print _("Aborting")
        usage()
        return 1     

    #else, execute the commands
    else:
        commands_to_execute = check_this_command(opt_command[0], template_path, opt_template)
        return_code = 0
        #pre-hook
        if 'pre' in commands_to_execute:
            return_code = commands_to_execute['pre'](opt_template, opt_command[1:])
        if return_code != 0:
            print _("ERROR: pre_%s command failed") % opt_command[0]
            print _("Aborting")
            return return_code

        #main execution
        if callable(commands_to_execute[opt_template]):
            return_code = commands_to_execute[opt_template](opt_template, opt_command[1:])
        else:
            return_code = subprocess.call(["python", commands_to_execute[opt_template], " ".join(opt_command[1:])])

        if return_code != 0:
            print _("ERROR: %s command failed") % opt_command[0]
            print _("Aborting")
            return return_code

        #post-hook
        if 'post' in commands_to_execute:
             commands_to_execute['post'](opt_template, opt_command[1:])
        if return_code != 0:
            print _("ERROR: post_%s command failed") % opt_command[0]
            print _("Aborting")
            return return_code

        return 0

if __name__ == '__main__':

    gettext.textdomain('quickly')

    #process the command line to send the right instructions
    exit(process_command_line())
